qiskit_optimization/translators/gurobipy.py

```diff
@@ -0,0 +1,280 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Model translators between QuadraticProgram and Gurobipy"""
+
+from typing import cast, Any
+
+from qiskit.exceptions import MissingOptionalLibraryError
+
+from qiskit_optimization.exceptions import QiskitOptimizationError
+from qiskit_optimization.problems.constraint import Constraint
+from qiskit_optimization.problems.quadratic_objective import QuadraticObjective
+from qiskit_optimization.problems.variable import Variable
+from .model_translator import ModelTranslator
+
+try:
+    import gurobipy as gp
+    from gurobipy import Model as GurobiModel
+
+    _HAS_GUROBI = True
+except ImportError:
+    _HAS_GUROBI = False
+
+    class GurobiModel:  # type: ignore
+        """ Empty GurobiModel class
+            Replacement if gurobipy.Model is not present.
+        """
+        pass
+
+
+class GurobipyTranslator(ModelTranslator):
+    """Translator between a gurobipy model and a quadratic program
+    """
+
+    def qp_to_model(self, quadratic_program: Any) -> GurobiModel:
```

Comment:
Perhaps `quadratic_program` is an instance of `QuadraticProgram`, not `Any`?

Reply:
It is impossible to avoid cyclic import by pylint if `quadratic_program: QuadraticProgram`. If we are allowed to disable pylint cyclic import, I can do it.

Reply:
Oh, I see the problem now.